syntax = "proto3";

enum RunStatus {
  PENDING = 0;
  RUNNING = 1;
  FAILED = 2;
  COMPLETED = 3;
  TERMINATED = 4;
}

message Obj_Project {
  string id = 1;
  int64 created_at = 2;
  int64 updated_at = 3;
  bool is_deleted = 4;
  string image_name = 5;
  string name = 6;
}

message Obj_Worker {
  enum Status {
    IDLE = 0;
    WORKING = 1;
    TERMINATING = 2;
    UNRESPONSIVE = 3;
    CLOSED = 4;
  }

  string id = 1;
  int64 created_at = 2;
  int64 updated_at = 3;
  bool is_deleted = 4;
  string project_id = 5;
  Status status = 6;
}

message Obj_Workflow {
  string id = 1;
  int64 created_at = 2;
  int64 updated_at = 3;
  bool is_deleted = 4;
  string name = 5;
  string project_ref_id = 6;
}

message Obj_WorkflowRun {
  string id = 1;
  int64 created_at = 2;
  int64 updated_at = 3;
  bool is_deleted = 4;
  bool job_id = 6;
  string project_ref_id = 7;
}

message Obj_WorkerLog {
  string id = 1;
  int64 created_at = 2;
  int64 updated_at = 3;
  bool is_deleted = 4;
  string payload = 5;
  string payload_key = 6;
  string worker_id = 7;
}

message Obj_WorkerDirective {
  string id = 1;
  int64 created_at = 2;
  int64 updated_at = 3;
  bool is_deleted = 4;
  string payload = 5;
  string payload_key = 6;
  string worker_receiver_id = 7;
  string worker_sender_id = 8;
}

message Obj_Task {
  string id = 1;
  int64 created_at = 2;
  int64 updated_at = 3;
  bool is_deleted = 4;
  bool is_mutable = 5;
  string name = 6;
  string project_ref_id = 7;
  string version = 8;
}

message Req_GetProject {
  string name = 1;
}

message Req_RegisterProject {
  string image_name = 1;
  string name = 2;
}

message Req_RegisterWorkflow {
  string name = 1;
  string project_ref_id = 2;
}

message Req_RegisterTask {
  string name = 1;
  string project_ref_id = 2;
  string version = 3;
}

message Req_RegisterWorker {
  string project_id = 1;
}

message Req_RouteWorkerDirective {
  string payload = 1;
  string payload_key = 2;
  string worker_id = 3;
}

message Res_RegisterWorker {
  string project_id = 1;
}

message Res_WriteWorkflowRunLog {
  enum Receipt {
    RECEIVED = 0;
    FAILED = 1;
  }

  string id = 1;
  Receipt receipt = 2; 
}

service ML {
  rpc GetProject(Req_GetProject) returns (Obj_Project);

  rpc RegisterProject(Req_RegisterProject) returns (Obj_Project);
  rpc RegisterWorker(Req_RegisterWorker) returns (Obj_Worker);
  rpc RegisterWorkflows(stream Req_RegisterWorkflow) returns (stream Obj_Workflow);
  rpc RegisterTasks(stream Req_RegisterTask) returns (stream Obj_Task);

  rpc RouteWorkerDirectives(stream Req_RouteWorkerDirective) returns (stream Obj_WorkerDirective);
}